File: swift-package/Sources/UniversalAiConnector/UniversalAiConnectorConfiguration.swift
/// Immutable provider-neutral construction data for one configured adapter.
///
/// The connector retains the supplier, not a credential returned by it. The
/// synchronous supplier is called once for each provider request and may read
/// host-owned secure storage at that time.
public struct UniversalAiProviderConfiguration: Sendable {
    public let providerId: UniversalAiProviderId
    public let baseURL: String
    internal let credentialSupplier: @Sendable () throws -> String

    public init(
        providerId: UniversalAiProviderId,
        baseURL: String,
        credentialSupplier: @escaping @Sendable () throws -> String
    ) {
        self.providerId = providerId
        self.baseURL = baseURL
        self.credentialSupplier = credentialSupplier
    }
}

/// Immutable per-client provider configuration.
public struct UniversalAiConnectorConfiguration: Sendable {
    public static let defaultConnectTimeoutMillis: Int64 = 10_000
    public static let defaultRequestTimeoutMillis: Int64 = 60_000
    public static let maximumTimeoutMillis: Int64 = 86_400_000

    public let providers: [UniversalAiProviderConfiguration]
    public let connectTimeoutMillis: Int64
    public let requestTimeoutMillis: Int64

    public init(
        providers: [UniversalAiProviderConfiguration] = [],
        connectTimeoutMillis: Int64 = Self.defaultConnectTimeoutMillis,
        requestTimeoutMillis: Int64 = Self.defaultRequestTimeoutMillis
    ) {
        self.providers = providers
        self.connectTimeoutMillis = connectTimeoutMillis
        self.requestTimeoutMillis = requestTimeoutMillis
    }
}

File: bridge/src/commonMain/kotlin/com/maneesh/universalai/connector/UniversalAiConnectorConfiguration.kt
@file:OptIn(kotlin.experimental.ExperimentalObjCRefinement::class)

package com.maneesh.universalai.connector

import com.maneesh.universalai.connector.contract.ProviderId
import com.maneesh.universalai.connector.internal.transport.ConnectorBaseUrl
import kotlin.native.HiddenFromObjC

/**
 * Immutable host-owned configuration for one canonical provider.
 *
 * The credential supplier is synchronous and is invoked once for each network request. The
 * connector retains the supplier, never a credential returned by it.
 */
@HiddenFromObjC
class UniversalAiProviderConfiguration(
    val providerId: ProviderId,
    baseUrl: String,
    credentialSupplier: () -> String,
) {
    internal val validatedBaseUrl: ConnectorBaseUrl = ConnectorBaseUrl.parse(baseUrl)
    internal val credentialSupplier: () -> String = credentialSupplier

    /**
     * The validated, normalized provider base URL.
     *
     * Provider traffic requires HTTPS. Plaintext HTTP is accepted only for exact loopback hosts
     * used by local development and test servers.
     */
    val baseUrl: String
        get() = validatedBaseUrl.value
}

/**
 * Immutable per-client provider configuration.
 *
 * Provider identifiers are unique. The empty configuration preserves the deterministic,
 * credential-free connector behavior.
 */
@HiddenFromObjC
class UniversalAiConnectorConfiguration(
    providers: List<UniversalAiProviderConfiguration>,
) {
    private val storedProviders: List<UniversalAiProviderConfiguration>

    init {
        val ordered = providers.sortedBy { provider -> provider.providerId.rawValue }
        val duplicate =
            ordered.zipWithNext().firstOrNull { (first, second) ->
                first.providerId == second.providerId
            }?.first
        require(duplicate == null) {
            "Provider '${duplicate?.providerId?.rawValue}' is configured more than once."
        }
        storedProviders = ordered.toList()
    }

    /** Returns a defensive snapshot in deterministic provider-identifier order. */
    val providers: List<UniversalAiProviderConfiguration>
        get() = storedProviders.toList()

    internal fun providersForRegistration(): List<UniversalAiProviderConfiguration> =
        storedProviders

    companion object {
        val Empty: UniversalAiConnectorConfiguration =
            UniversalAiConnectorConfiguration(emptyList())
    }
}
